Build failure: Pymol

## Steps To Reproduce

Steps to reproduce the behavior:

1. build pymol

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
error: builder for '/nix/store/rlvzvpdmzhglm2cwcv0j42sdji50p768-pymol-3.0.0.drv' failed with exit code 1;
       last 25 log lines:
       >                 raise AttributeError("`numpy.distutils` is not available from "
       >                                      "Python 3.12 onwards", name=None)
       >
       >         if attr in __future_scalars__:
       >             # And future warnings for those that will change, but also give
       >             # the AttributeError
       >             warnings.warn(
       >                 f"In the future `np.{attr}` will be defined as the "
       >                 "corresponding NumPy scalar.", FutureWarning, stacklevel=2)
       >
       >         if attr in __former_attrs__:
       >             raise AttributeError(__former_attrs__[attr], name=None)
       >
       >         if attr in __expired_attributes__:
       > >           raise AttributeError(
       >                 f"`np.{attr}` was removed in the NumPy 2.0 release. "
       >                 f"{__expired_attributes__[attr]}",
       >                 name=None
       >             )
       > E           AttributeError: `np.asfarray` was removed in the NumPy 2.0 release. Use `np.asarray` with a proper dtype instead.
       >
       > /nix/store/8y0ibahym22givsbmz5z636gwxw1k7zk-python3.11-numpy-2.2.0/lib/python3.11/site-packages/numpy/__init__.py:400: AttributeError
       > =========================== short test summary info ============================
       > FAILED tests/api/chempybrick.py::TestChempyBrick::test - AttributeError: `np.asfarray` was removed in the NumPy 2.0 release. Use `np...
       > ========== 1 failed, 646 passed, 326 skipped, 8 deselected in 32.41s ===========
       For full logs, run 'nix log /nix/store/rlvzvpdmzhglm2cwcv0j42sdji50p768-pymol-3.0.0.drv'.
error: 1 dependencies of derivation '/nix/store/ymk30cfzs5dr4j54xh1ws0znzvarrxq4-nix-shell-env.drv' failed to build
```

</details>

## Additional context

<!-- Add any other context about the problem here. -->
using latest nixos-unstable commit(devshell, waiting for my other issues to be fixed b4 updating system input)

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
- system: `"x86_64-linux"`
 - host os: `Linux 6.12.5, NixOS, 25.05 (Warbler), 25.05.20241219.d70bd19`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - channels(root): `"nixos-24.05"`
 - nixpkgs: `/nix/store/8fwsiv0hd7nw1brkvka0jf1frk3m7qkr-source`

## Notify maintainers
@tomoyat1 @samlich 
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/applications/science/chemistry/pymol/default.nix
+++ b/pkgs/applications/science/chemistry/pymol/default.nix
@@ -3,6 +3,7 @@
   lib,
   fetchFromGitHub,
   makeDesktopItem,
+  cmake,
   python3Packages,
   netcdf,
   glew,
@@ -10,6 +11,7 @@
   libpng,
   libxml2,
   freetype,
+  mmtf-cpp,
   msgpack,
   qt5,
 }:
@@ -43,37 +45,51 @@ let
 in
 python3Packages.buildPythonApplication rec {
   inherit pname;
-  version = "3.0.0";
+  version = "3.1.0";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "schrodinger";
     repo = "pymol-open-source";
     rev = "v${version}";
-    hash = "sha256-GhTHxacjGN7XklZ6gileBMRZAGq4Pp4JknNL+qGqrVE=";
+    hash = "sha256-2C9kUpNfK9g7ehmk83iUVqqz4gn4wKO3lW5rSduFP6U=";
   };
+
+  # A script is already created by the `[project.scripts]` directive
+  # in `pyproject.toml`.
+  patches = [ ./script-already-exists.patch ];
 
   postPatch = ''
     substituteInPlace setup.py \
       --replace-fail "self.install_libbase" '"${placeholder "out"}/${python3Packages.python.sitePackages}"'
   '';
 
-  build-system = [
-    python3Packages.setuptools
+  env.PREFIX_PATH = lib.optionalString (!stdenv.hostPlatform.isDarwin) "${msgpack}";
+  build-system = [ python3Packages.setuptools ];
+  dontUseCmakeConfigure = true;
+
+  nativeBuildInputs = [
+    cmake
+    qt5.wrapQtAppsHook
   ];
 
-  nativeBuildInputs = [ qt5.wrapQtAppsHook ];
-  buildInputs = [
-    python3Packages.numpy
-    python3Packages.pyqt5
-    glew
-    glm
-    libpng
-    libxml2
-    freetype
-    msgpack
-    netcdf
-  ];
+  buildInputs =
+    [
+      python3Packages.numpy_1
+      python3Packages.pyqt5
+      qt5.qtbase
+      glew
+      glm
+      libpng
+      libxml2
+      freetype
+      netcdf
+    ]
+    ++ lib.optionals (!stdenv.hostPlatform.isDarwin) [
+      mmtf-cpp
+      msgpack
+    ];
+
   env.NIX_CFLAGS_COMPILE = "-I ${libxml2.dev}/include/libxml2";
 
   postInstall =
@@ -94,9 +110,7 @@ python3Packages.buildPythonApplication rec {
       cp -r "${desktopItem}/share/applications/" "$out/share/"
     '';
 
-  pythonImportsCheck = [
-    "pymol"
-  ];
+  pythonImportsCheck = [ "pymol" ];
 
   nativeCheckInputs = with python3Packages; [
     python3Packages.msgpack
@@ -112,17 +126,22 @@ python3Packages.buildPythonApplication rec {
     "tests/api/seqalign.py"
   ];
 
-  disabledTests = [
-    # the output image does not exactly match
-    "test_commands"
-    # touch the network
-    "testFetch"
-    # requires collada2gltf which is not included in nixpkgs
-    "testglTF"
-    # require mmtf-cpp which does not support darwin
-    "testMMTF"
-    "testSave_symmetry__mmtf"
-  ];
+  disabledTests =
+    [
+      # the output image does not exactly match
+      "test_commands"
+      # touch the network
+      "testFetch"
+      # requires collada2gltf which is not included in nixpkgs
+      "testglTF"
+    ]
+    ++ lib.optionals (stdenv.hostPlatform.isDarwin) [
+      # require mmtf-cpp which does not support darwin
+      "test_bcif"
+      "test_bcif_array"
+      "testMMTF"
+      "testSave_symmetry__mmtf"
+    ];
 
   preCheck = ''
     cd testing

--- a/pkgs/applications/science/chemistry/pymol/script-already-exists.patch
+++ b/pkgs/applications/science/chemistry/pymol/script-already-exists.patch
@@ -1,0 +1,23 @@
+diff --git a/setup.py b/setup.py
+index 507dd32d..bf258897 100644
+--- a/setup.py
++++ b/setup.py
+@@ -435,9 +435,6 @@ class install_pymol(install):
+         super().run()
+         self.install_pymol_path()
+ 
+-        if not self.no_launcher:
+-            self.make_launch_script()
+-
+         if self.bundled_pmw:
+             raise Exception(
+                 "--bundled-pmw has been removed, please install Pmw from "
+@@ -487,7 +484,7 @@ class install_pymol(install):
+         launch_script = os.path.join(self.install_scripts, launch_script)
+ 
+         python_exe = os.path.abspath(sys.executable)
+-        site_packages_dir = sysconfig.get_path('purelib')
++        site_packages_dir = sysconfig.get_path("purelib")
+         pymol_file = self.unchroot(
+             os.path.join(site_packages_dir, "pymol", "__init__.py")
+         )

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -16774,9 +16774,7 @@ with pkgs;
 
   pdb2pqr = with python3Packages; toPythonApplication pdb2pqr;
 
-  pymol = callPackage ../applications/science/chemistry/pymol {
-    python3Packages = python311Packages;
-  };
+  pymol = callPackage ../applications/science/chemistry/pymol { };
 
   quantum-espresso = callPackage ../applications/science/chemistry/quantum-espresso {
     hdf5 = hdf5-fortran;

